### ISSUE TITLE
Docker support

### ISSUE THREAD

**Issue by kylerw** [mdhiggins/sickbeard_mp4_automator#412]
I run all my apps in separate docker containers and would love for the ability to do the same here.

Not sure of the best method, but possibly exposing an API of sorts to trigger the script externally (since each container are essentially silo-ed but have access to the same file structure).

**Comment by mdhiggins** [comment]
Never used docker and not really sure what it offers, but I'd imagine this would be challenging to implement

Probably not something I'll have time to get to

**Comment by kylerw** [comment]
@mdhiggins I think all the would be needed is a service that could be called to execute or start scripts. Think of it as the services on a separate computer from mp4automator. If there was a service/app that could be called it would facilitate dockwr capabilities to be built.

**Comment by KaHooli** [comment]
I've to been looking into this, and I'm only very new to Docker as well.

Docker allows you to run applications in a sandboxed environment called containers (think shipping containers on a cargo ship).  That is, the host is a linux OS with Docker installed, then you run each application in their own container.  Each container can then have the dependencies they required installed above what the host OS already has installed, without effecting other containers or the host OS.  This in turn allows the distribution of containers very easily, as the docker container distribution file is only a script stating where to source the different components from and how to install them.

This especially interests me as Synology supports Docker on their NAS OS (called DSM), and running this within a docker container would mean less modifications of the host's OS.

I can see some people have stated incorporating sickbeard_mp4_automator scripts into docker containers already...
- [Example that includes this script, FFmpeg, Python, SB and CP](https://github.com/remlabm/docker-sickbeard-mp4)
- [Another that includes this script, FFmpeg, Python, Sonarr and CP](https://github.com/milleruk/docker-sonarr-mp4)

What I'm looking at doing is...

1. [Sonarr](https://synocommunity.com/package/nzbdrone) or [CouchPotato](https://synocommunity.com/package/couchpotatoserver) (both installed as a Synology packages) finds the desired item and sends it to...
2. [NZBGet](https://synocommunity.com/package/nzbget) (again, installed as a Synology package), which downloads the desired item and executes a NZBGet script that...
3. runs the [sickbeard_mp4_automator](https://github.com/mdhiggins/sickbeard_mp4_automator) script which is inside a docker container.  The script has the NZBGet download directory mapped as /NZBGet within the container, converts the item and passes console output back through to NZBGet script
4. Once conversion is done, the NZBGet script advises Sonarr or CouchPotato so they can move the file to the right directory

As I'm new to Docker, this is taking awhile to do, and I'm currently working to get step 3 above running manually first, then I'll look at how it can be automated.  The Docker container image script needs to be able to check and install if required
- Python
 - and all dependencies
- compile FFmpeg specific to the host's CPU type
- map a folder to the download clients download directory
- install the sickbeard_mp4_automator script.
- map the autoProcess.ini file to a location outside of the container, so settings aren't lost between updates

Anyway, I know this is a long post, and I hope it helps.

**Comment by kylerw** [comment]
Ended up creating a script that I run inside if my NZBGet container that installs MP4A. Check my repos and let me know if you have questions.

**Comment by StevenChorkley** [comment]
How do you automatically call your script when creating a new container once the image has been updated?

**Comment by kylerw** [comment]
@StevenChorkley something like this: 

docker exec NZBGet sh /config/userscript.sh


That can be thrown into a script and run using cron or manually each update.

**Comment by tac00** [comment]
@kylerw is it possible to create a script like this just for Radarr and Sonarr?? I also run my software in containers and will like to use both these with MP4 automator script.

**Comment by kylerw** [comment]
The script should be the same, given that Radarr and Sonarr use a similar base image (I'm using LinuxServer versions and they all do). I found that I only needed a single version installed (Downloader makes the most sense) and the file would be converted then passed to Sonarr/Radarr/etc. So the script shouldn't need to be run in each container...

**Comment by tac00** [comment]
@kylerw This is great news. I use the same docker images all from Linuxserver.io. Im not sure I understand the last part correctly. Im using Deluge, Sonarr and Radar. What container do I need to run the script in?

**Comment by StevenChorkley** [comment]
Description=Radarr container
Requires=docker.service
After=docker.service

[Service]
Restart=always
RestartSec=30
ExecStart=/usr/bin/docker run --rm --name radarr -p 7878:7878 -e PUID=1000 -e PGID=1000 -e TZ=<timezone>
ExecStartPost=/usr/bin/docker exec NZBGet sh /config/userscript.sh
ExecStop=/usr/bin/docker stop radarr

[Install]
WantedBy=multi-user.target

**Comment by kylerw** [comment]
@StevenChorkley that looks like it should be correct. I never played with Sonarr/Radarr tagging and only ever had mp4a running in my NZBGet container. I'm not sure of all the nuances, but you could probably create a similar script for sonarr/radarr to perform the tagging there. Downside, is you then have it installed 3 times and they are all separate instances.... there may be a better way.

**Comment by StevenChorkley** [comment]
Is it possible to build it into the docker file? Might be able to do the same for sonarr and Radarr

**Comment by StevenChorkley** [comment]
Should be okay though as they're called separately anyway. The conversion would happen in the NZBGet container and the tagging in sonarr/radarr

**Comment by kylerw** [comment]
@StevenChorkley yes, if you create/modify the dockerfile you should be able to add all dependencies and install mp4a without any issues.

**Comment by StevenChorkley** [comment]
Thanks for your help! 
My knowledge of docker is very poor. Have you done anything similar I can look at?

**Comment by kylerw** [comment]
@StevenChorkley yeah, that's probably the best route.

check out: https://github.com/kylerw/appdaemon-docker

basically, you'd want to fork the docker container and update the dockerfile/install so that it added all the mp4a dependencies.

**Comment by tac00** [comment]
I will fork the Deluge docker container by linuxserver and add my script that installs mp4a and dependencies when I get home and have the time. I have tested it and it works fine with both Radarr and Sonarr.

@StevenChorkley if you want to run your software in docker you can install my image and use Deluge as downloader and Radarr and Sonarr as media managers. All software in their own container.

**Comment by KaHooli** [comment]
I've had to put this on hold for awhile now, but was thinking about this just now...

I'm finding that this script needs to be run in the same container as the Downloader (eg NZBget) as its the downloader that calls the script.

The problem I have is when it tells the Media Manager (eg Sonarr or CouchPotato) where the file is located, as the file path to the downloaded file can be different different between the host, and individual containers.

It would be good to have an option for each Media Manager to convert the path for the Downloader to the path for the Media Manager, example...
```
[Sonarr]
host = 192.168.0.2
port = 8989
web_root =
ssl = False
apikey = blah1blah2blah3
remotepath = /media/download/,/volume1/Public/Downloads/NZBGet/
```
So if the value is left blank, no conversion needed, but if it is like the example above then when the script passes the path to the Download Manager, it doesn't error.

What do you think?

**Comment by kylerw** [comment]
@KaHooli that's a good point, I have my containers mapped identical. So the path for the downloader translates for the manager without any extra settings.

**Comment by tac00** [comment]
@kylerw Im curious how you handle compressed content with your setup? Im lookin into using Filebot as my renaming/moving tool and then call mp4a to convert after renaming by Filebot. Filebot can handle unraring of compressed data. Radarr, Sonarr, mp4a cant handle that yet :(

**Comment by StevenChorkley** [comment]
@tac00 cheers, let me know when you publish!

**Comment by KaHooli** [comment]
As the scripts as dependencies, I don't believe it will work in a seperate container.

I've forked the linuxserver/nzbget to include the some scripts, but it is not close to what I would call polished.  https://hub.docker.com/r/kahooli/nzbget-postprocess/

**Comment by StevenChorkley** [comment]
Awesome! Any plans to get tagging working with Sonarr / Radarr? Any reason as to why you've used both scripts?

**Comment by KaHooli** [comment]
I use this script, but I know people who use nzbToMedia. The nzbToMedia script isn't fully functional yet in this docker image.
Tagging seems to be working for me.

**Comment by StevenChorkley** [comment]
Awesome, how did you get the postSonarr.py and postRadarr.py setup in the Sonarr and Radarr docker images?

**Comment by StevenChorkley** [comment]
I can't get your dockerfile to work, I only have nzbget.conf in my /config directory. Whenever I try to connect using bash, there's no output from the container. 

NZBGet is running, but the MP4 script location isn't set. 

I'm at a complete loss.

**Comment by niallobr** [comment]
This is definitely possible without any additional scripting!

I've just finished setting up separate linuxserver containers for Deluge, SABnzbd, Sonarr, Radarr and Jackett on my Synology NAS.

sickbeard_mp4_automator is running locally and handles all the conversions sent from these containers. I'm still testing but everything seems to be working well so far.

The main requirements seem to be:

1. Mounting your local storage within the containers in a way that matches the local file structure allowing these to all be visible/writable/executable:

• Your downloads folders
• Your media folders
• ffmpeg and ffprobe

2. Matching uid's/gid's between the two (not sure if this is required but I did it anyway)

3. Setting the correct permissions

4. Installing dependencies in the containers

5. Specifying your host IP addresses in autoProcess.ini instead of localhost (because the containers are bridged to the host)

I think that's it, I'll update if I think of anything else.. It's not really that complicated to setup and permissions were probably the biggest headache.

I'm hoping to post a more detailed guide on my blog.

**Comment by StevenChorkley** [comment]
How do you execute the script on the host from the container?

**Comment by niallobr** [comment]
@StevenChorkley Hey Steven,

I have 2 local volumes mounted in each container. So for example in your root directory in each container you might have **/myvolumes/downloads** and **/myvolumes/video/**.

I have my folders mounted as /volume1/downloads and /volume1/video because it matches the folder structure on my Synology host.

I placed the mp4_converter script inside the video folder. This is accessible by a users group which includes all the programs I'm using (Sonarr, Deluge etc..).

I have the SABnzbd scripts folder pointed to /volume1/video/mp4converter and this executes SABPostProcess.py as expected. Similarly with Deluge the Execute plugin points to delugePostProcess.py.

There's a bit of checking to make sure all the permissions are correct and dependencies are present in order for it to work, I did a bit of trial and error to get all that figured out!

I'll create a map to explain this a bit clearer and will post a link when I have time. I'm far from a Linux expert so I think most should be able to get it up and running with some time and effort!

**Comment by StevenChorkley** [comment]
So you executed the python scripts stored on the host inside the docker containers? @niallobr

**Comment by haxcop** [comment]
@tac00 Hey man have you found a way to work with filebot into the autoprocess script?

**Comment by mdhiggins** [comment]
If you have some info regarding filebot's post processing features I can see about integration, point me in the right direction

**Comment by mdhiggins** [comment]
So I just recently rebuilt my plex server and did a debian / docker setup

Ended up creating my own dockerfiles based on the linuxserver ones for radarr and sonarr and am using those with completed download handling

Figured I'd share here and see if anyone else has some recommendations or wants to share their containers

https://hub.docker.com/r/mdhiggins/sonarr-sma/
https://hub.docker.com/r/mdhiggins/radarr-sma/

**Comment by letchu22** [comment]
Just tried the docker image on Unraid. Been using this on Windows for some time. Works great! Not sure why I can't get it working on Unraid. Sonarr portion woks fine. Might be careless somewhere or not getting you correctly. This is generally what i did. 

Volumes:
1.  /mnt/user/appdata/sonarr-sma:/config  
2. /mnt/user/appdata/sma:/config/sma - set as read only
3. /mnt/user/appdata/sma/autoProcess.ini:/usr/local/bin/sma/sickbeard_mp4_automator/autoProcess.ini - set as read only
4. /mnt/user/media/TV Shows:/tv
5. /mnt/user/downloads/completed:/downloads

Sonarr:
1. Completed Download Handling: On
2. Custom Script: path = /usr/local/bin/sma/sickbeard_mp4_automator/postSonarr.py - test is successful
3. autoProcess.ini visible from both /config/sma and /usr/local/bin/sma/sickbeard_mp4_automator

Autoprocess.ini
1. ffmpeg = /usr/local/bin/ffmpeg/ffmpeg/ffmpeg.exe
2. ffprobe = /usr/local/bin/ffmpeg/ffprobe/ffprobe.exe
3. Sonarr portion is filled up with appropriate values

Not too sure how to access the log to submit.  Any help towards where to look will be appreciated.

**Comment by mdhiggins** [comment]
Ah I bet I know what's going on

The image sandboxes the python installation into a venv within the docker and you need to tell Sonarr to use that python environment

Here's what mine looks like

![image](https://user-images.githubusercontent.com/3608298/43839695-bdb7de08-9aec-11e8-9af6-efcf7fae67c5.png)

Path: `/usr/local/bin/sma/env/bin/python3` 
Arg: `/usr/local/bin/sma/sickbeard_mp4_automator/postSonarr.py`

I'll need to update the readme for these

And just in case its helpful, here's my full docker-compose.yml for my server
```
# /opt/docker-compose.yml
---
version: '3.6'
services:
  plex:
    image: plexinc/pms-docker:public
    container_name: plex
    volumes:
      - /opt/appdata/plex:/config
      - /mnt/storage:/data
      - /etc/localtime:/etc/localtime:ro
    network_mode: host
    restart: always
    environment:
      - PLEX_UID=${PUID}
      - PLEX_GID=${PGID}
      - PLEX_CLAIM=${PLEX_CLAIM}
  tautulli:
    image: linuxserver/tautulli
    container_name: tautulli
    volumes:
      - /opt/appdata/tautulli:/config
      - /opt/appdata/plex/Library/Application\ Support/Plex\ Media\ Server/Logs:/logs:ro
      - /etc/localtime:/etc/localtime:ro
    ports:
      - 8181:8181
    restart: always
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
  nordvpn:
    image: bubuntux/nordvpn
    container_name: nordvpn
    environment:
      - USER=${NORDUSER}
      - PASS=${NORDPASS}
      - NETWORK=10.0.1.0/24
    ports:
      - 6789:6789 # NZBGet
      - 8989:8989 # Sonarr
      - 7878:7878 # Radarr
      - 8112:8112 # Deluge
    devices: 
      - /dev/net/tun
    cap_add:
      - NET_ADMIN
    restart: always
  nzbget:
    image: linuxserver/nzbget
    container_name: nzbget
    depends_on:
      - nordvpn
    network_mode: service:nordvpn
    volumes:
      - /opt/appdata/nzbget/config:/config
      - /mnt/storage/downloads:/downloads
      - /etc/localtime:/etc/localtime:ro
    restart: always
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
  sonarr:
    image: mdhiggins/sonarr-sma
    container_name: sonarr
    depends_on:
      - nordvpn
    network_mode: service:nordvpn
    volumes:
      - /opt/appdata/sonarr:/config
      - /opt/appdata/sma/autoProcess.ini:/usr/local/bin/sma/sickbeard_mp4_automator/autoProcess.ini:ro
      - /mnt/storage/tv:/tv
      - /mnt/storage/downloads:/downloads
      - /etc/localtime:/etc/localtime:ro
    restart: always
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
  radarr:
    image: mdhiggins/radarr-sma
    container_name: radarr
    depends_on:
      - nordvpn
      - nzbget
      - deluge
    network_mode: service:nordvpn
    volumes:
      - /opt/appdata/radarr:/config
      - /opt/appdata/sma/autoProcess.ini:/usr/local/bin/sma/sickbeard_mp4_automator/autoProcess.ini:ro
      - /mnt/storage/movies:/movies
      - /mnt/storage/downloads:/downloads
      - /etc/localtime:/etc/localtime:ro
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
    restart:
      always
  deluge:
    image: linuxserver/deluge
    container_name: deluge
    depends_on:
      - nordvpn
    network_mode: service:nordvpn
    volumes:
      - /opt/appdata/deluge:/config
      - /mnt/storage/downloads:/downloads
      - /mnt/storage:/mnt/storage
    restart: always
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
  smartmontools:
    image: mdhiggins/smartmontools
    privileged: true
    container_name: smartmontools
    environment:
      - MAILTO=${MAILTO}
      #- MAILFROM=${MAILFROM}
      #- MAILPASS=${MAILPASS}
      #- SMTP=${SMTP}
    volumes:
      - /opt/appdata/smartmontools/smartd.conf:/etc/smartd.conf:ro
      - /opt/appdata/smartmontools/ssmtp/ssmtp.conf:/etc/ssmtp/ssmtp.conf:ro
    restart: always
  snapraid:
    image: xagaba/snapraid
    privileged: true
    container_name: snapraid
    environment:
      - PUID=${PUID}
      - PGID=${PGID}
    volumes:
      - /mnt:/mnt
      - /var/snapraid:/var/snapraid
      - /opt/appdata/snapraid:/config:ro
      - /etc/localtime:/etc/localtime:ro  
      - type: bind
        source: /dev/disk
        target: /dev/disk
```

I have most images that do any downloading routed through a nordVPN container
Also have smartmontools which will alert via email of HDD failure, and snapraid which handles the parity and backup

**Comment by totallyGreg** [comment]
I'm happy to see you're providing sonarr/radarr images with sma, although I am curious as to why not package up a single sma image and use it as a sidecar to existing images?  
I have been playing with creating a Dockerfile to do that when I noticed this thread.  

In the meantime testing out your mdhiggins/radarr-sma image I noticed readSettings.py is expecting the autoProcess.ini to be a writable file.  

```
root@radarr-sma-0:/usr/local/bin/sma/sickbeard_mp4_automator# /usr/local/bin/sma/env/bin/python3 manual.py -a /tmp/bobs.burgers.s08e18.1080p.web.x264-tbs.mkv
Manual processor started.
/usr/local/bin/sma/env/bin/python3
Traceback (most recent call last):
  File "manual.py", line 42, in <module>
    settings = ReadSettings(os.path.dirname(sys.argv[0]), "autoProcess.ini", logger=log)
  File "/usr/local/bin/sma/sickbeard_mp4_automator/readSettings.py", line 210, in __init__
    self.writeConfig(config, configFile)
  File "/usr/local/bin/sma/sickbeard_mp4_automator/readSettings.py", line 695, in writeConfig
    fp = open(cfgfile, "w")
OSError: [Errno 30] Read-only file system: 'autoProcess.ini'
```

**Comment by mdhiggins** [comment]
Hm looks like that was an oversight the script should have write permissions because it will add options to the config file if they are missing (like if a new option is added in a future update)

And creating separate docker images based on the linuxserver ones just seemed like the simplest solution to get the most benefit out of docker. I haven't done much with sidecar services and this is my first attempt at ever using docker.

**Comment by NEprod** [comment]
sorry to but in. I've been trying to get this to work followed all steps above I've seen etc. using your own version get this error same as above I would say is there away I could fix this?

python3 | PermissionError: [Errno 13] Permission denied: '/usr/local/bin/sma/sickbeard_mp4_automator/autoProcess.ini' | Yesterday
-- | -- | --
  | python3 | fp = open(cfgfile, "w") | Yesterday
  | python3 | File "/usr/local/bin/sma/sickbeard_mp4_automator/readSettings.py", line 695, in writeConfig | Yesterday
  | python3 | self.writeConfig(config, configFile) | Yesterday
  | python3 | File "/usr/local/bin/sma/sickbeard_mp4_automator/readSettings.py", line 210, in __init__ | Yesterday
  | python3 | settings = ReadSettings(os.path.dirname(sys.argv[0]), "autoProcess.ini") | Yesterday
  | python3 | File "/usr/local/bin/sma/sickbeard_mp4_automator/postRadarr.py", line 28, in <module> | Yesterday
  | python3 | Traceback (most recent call last):

**Comment by mdhiggins** [comment]
You need to give write permission to autoProcess.ini

**Comment by ShiroiKuma** [comment]
It's very nice of you to supply a docker-compose.yml that uses the docker images.

I'm not sure what the purpose of /mnt/storage:/mnt/storage on deluge is though, as that's not referenced in any other of the containers.

Anyway my question is, if one used this docker-compose.yml, what would the recommended autoProcess.ini look like? I'm using mostly defaults and leaving output_directory/move_to/copy_to blank.

**Comment by mdhiggins** [comment]
Hm looks like the deluge thing was an oversight, it doesn't need the full /mnt/storage access

And yes I'm using completed download handling instead of the post(deluge/nzbget) scripts in these containers
